Marlin/src/module/ft_motion.cpp

```diff
@@ -541,259 +272,277 @@ void FTMotion::setTrajectoryType(const TrajectoryType type) {
+    // Cache the extruder index for this block
+    TERN_(DISTINCT_E_FACTORS, block_extruder_axis = E_AXIS_N(current_block->extruder));
 
-  // Accel + Coasting + Decel + datapoints
-  const float reminder_from_last_block = - tau;
+    const float totalLength = current_block->millimeters;
 
-  endPos_prevBlock += moveDist;
+    startPos = endPos_prevBlock;
+    const xyze_pos_t& moveDist = current_block->dist_mm;
+    ratio = moveDist / totalLength;
 
-  TERN_(FTM_HAS_LIN_ADVANCE, use_advance_lead = current_block->use_advance_lead);
+    const float mmps = totalLength / current_block->step_event_count, // (mm/step) Distance for each step
+                initial_speed = mmps * current_block->initial_rate,   // (mm/s) Start feedrate
+                final_speed = mmps * current_block->final_rate;       // (mm/s) End feedrate
 
-  // Watch endstops until the move ends
-  const float total_duration = currentGenerator->getTotalDuration();
-  uint32_t max_intervals = ceil((total_duration + reminder_from_last_block) * FTM_FS);
-  const millis_t move_end_ti = millis() + SEC_TO_MS((FTM_TS) * float(max_intervals + num_samples_shaper_settle() + ((PROP_BATCHES) + 1) * (FTM_BATCH_SIZE)) + (float(FTM_STEPPERCMD_BUFF_SIZE) / float(FTM_STEPPER_FS)));
+    // Plan the trajectory using the trajectory generator
+    currentGenerator->plan(initial_speed, final_speed, current_block->acceleration, current_block->nominal_speed, totalLength);
 
-  #define _SET_MOVE_END(A) do{ \
-    if (moveDist.A) { \
-      axis_move_end_ti.A = move_end_ti; \
-      axis_move_dir.A = moveDist.A > 0; \
-    } \
-  }while(0);
+    endPos_prevBlock += moveDist;
 
-  LOGICAL_AXIS_MAP(_SET_MOVE_END);
-}
+    TERN_(FTM_HAS_LIN_ADVANCE, use_advance_lead = current_block->use_advance_lead);
 
-// Generate data points of the trajectory.
-// Called from FTMotion::loop() at the fetch of a new planner block, after loadBlockData.
-void FTMotion::generateTrajectoryPointsFromBlock() {
-  const float total_duration = currentGenerator->getTotalDuration();
-  if (tau + FTM_TS > total_duration) {
-    // TODO: refactor code so this thing is not twice.
-    // the reason of it being in the beginning, is that a block can be so short that it has
-    // zero trajectories.
-    // the next iteration will fall beyond this block
-    blockProcRdy = false;
-    traj_idx_get = 0;
-    tau -= total_duration;
-    return;
+    // Watch endstops until the move ends
+    const millis_t move_end_ti = millis() + \
+      stepper_plan_count() * FTM_TS + // Time to empty stepper command buffer
+      SEC_TO_MS(currentGenerator->getTotalDuration()) + // Time to finish this block
+      SEC_TO_MS((FTM_TS) * calc_runout_samples()); // Time for a rounout block
+
+    #define _SET_MOVE_END(A) do{ \
+      if (moveDist.A) { \
+        axis_move_end_ti.A = move_end_ti; \
+        axis_move_dir.A = moveDist.A > 0; \
+      } \
+    }while(0);
+
+    LOGICAL_AXIS_MAP(_SET_MOVE_END);
```

Comment:
@thinkyhead , you can remove PR #28125 and perhaps we could only keep `axis_move_dir = current_block->direction_bits`.
```suggestion
   axis_move_dir = current_block->direction_bits;
   #define _SET_MOVE_END(A) do{ \
     if (moveDist.A) { \
       axis_move_end_ti.A = move_end_ti; \
     } \
   }while(0);
   
    LOGICAL_AXIS_MAP(_SET_MOVE_END);
```

Reply:
I think we should revise the axis move ti thing completely (not within this pr). Now there is a single buffer so this whole time math is most likely not even needed. We could just set the last moved buffer index when a block is loaded, and clear it when that index is dequeued.

Reply:
Actually we could do that from within the queue and dequeue functions. Axis that move have a non zero dividend

Reply:
It's just temporary unless this can be refactored. I've found a solution in current bugfix but i always loose a ftmotion_stepper() call.

Reply:
Interesting, for me (with the other probe patch on top) it works as is. Should I apply your patch as is or does it require extra stuff?

Reply:
For axis_move_dir only the diff above. In the current bugfix, I create an empty command which marks the start of the block. When losing one ftmotion_stepper() call, everything is fine but not when I try to process the next command if the buffer is not empty.

Reply:
The main thing is to make sure the "is moving" and "current direction" are related to the immediate state, where "is moving" refers to the block currently being stepped.

Strictly speaking, all an endstop check wants to know is, "is the step count on the axis closer to the endstop end than it was the last time I checked?" and the same goes for probes wanting to check if the new position is lower.

There isn't much besides endstop checking that cares about having these states prepared, so if it is possible to refactor endstops and probes to use these counts instead, that would free up a few cycles from the motion code and localize more of the endstop / probe logic in their respective classes.